File: python/src/reality_capture/specifications/reconstruction.py
from pydantic import BaseModel, Field
from typing import Optional, Union
from tiling import ReferenceModel, TilingOptions, TilingOutputsCreate
from production import Export, ExportCreate, ProductionOptions


class ReconstructionInputs(BaseModel):
    scene: str = Field(description="Reality data id of ContextScene to process")
    region_of_interest: Optional[str] = Field(description="Path to region of interest file prefix by reality data id",
                                              alias="regionOfInterest", default=None)
    reference_model: Optional[str] = Field(None, description="Reality data id of reference model to process",
                                           alias="referenceModel")
    preset: Optional[str] = Field(default=None, description="Path to preset")


class ReconstructionOutputs(BaseModel):
    reference_model: Optional[ReferenceModel] = Field(None, description="Reference Model", alias="referenceModel")
    exports: Optional[list[Export]] = Field(description="List of exports")


class ReconstructionOptions(TilingOptions, ProductionOptions):
    pass


class ReconstructionSpecificationsCreate(BaseModel):
    inputs: ReconstructionInputs = Field(description="Inputs")
    outputs: list[Union[TilingOutputsCreate, ExportCreate]] = Field(description="Outputs")
    options: ReconstructionOptions = Field(description="Options")


class ReconstructionSpecifications(BaseModel):
    inputs: ReconstructionInputs = Field(description="Inputs")
    outputs: ReconstructionOutputs = Field(description="Outputs")
    options: ReconstructionOptions = Field(description="Options")

File: typescript/packages/reality-capture/src/specifications/reconstruction.ts
import { z } from "zod";
import { ModelingReferenceSchema, TilingOptionsSchema } from "./tiling";
import { ExportSchema, ExportCreateSchema } from "./production";

export const ReconstructionInputsSchema = z.object({
  scene: z.string().describe("Reality data id of ContextScene to process"),
  regionOfInterest: z
    .string()
    .regex(/^bkt:.+/)
    .describe(
      "Path in the bucket to region of interest file, used for tiling region of interest"
    )
    .optional(),
  extent: z
    .string()
    .regex(/^bkt:.+/)
    .describe(
      "Path in the bucket to region of interest file, used for export extent"
    )
    .optional(),
  modelingReference: z
    .string()
    .describe("Reality data id of modeling reference to process")
    .optional(),
  presets: z
    .array(z.string())
    .describe("List of paths to preset")
    .optional(),
  crsData: z.string().regex(/^bkt:.+/).optional().describe("Path in the bucket for CRS data."),
});
export type ReconstructionInputs = z.infer<typeof ReconstructionInputsSchema>;

export const ReconstructionOutputsSchema = z.object({
  modelingReference: ModelingReferenceSchema
    .optional()
    .describe("Modeling reference"),
  exports: z
    .array(ExportSchema)
    .optional()
    .describe("List of exports")
});
export type ReconstructionOutputs = z.infer<typeof ReconstructionOutputsSchema>;

export const ReconstructionOutputsCreateSchema = z.object({
  modelingReference: z
    .boolean()
    .optional()
    .describe("Modeling reference"),
  exports: z
    .array(ExportCreateSchema)
    .optional()
    .describe("Exports")
});
export type ReconstructionOutputsCreate = z.infer<typeof ReconstructionOutputsCreateSchema>;

export const ReconstructionSpecificationsCreateSchema = z.object({
  inputs: ReconstructionInputsSchema.describe("Inputs"),
  outputs: ReconstructionOutputsCreateSchema.describe("Outputs"),
  options: TilingOptionsSchema
    .optional()
    .describe("Options")
});
export type ReconstructionSpecificationsCreate = z.infer<typeof ReconstructionSpecificationsCreateSchema>;

export const ReconstructionSpecificationsSchema = z.object({
  inputs: ReconstructionInputsSchema.describe("Inputs"),
  outputs: ReconstructionOutputsSchema.describe("Outputs"),
  options: TilingOptionsSchema
    .optional()
    .describe("Options")
});
export type ReconstructionSpecifications = z.infer<typeof ReconstructionSpecificationsSchema>;

export const ReconstructionCostSchema = z.object({
  gpix: z
    .number()
    .min(0)
    .describe("Number of GigaPixels in the overall inputs."),
  mpoints: z
    .number()
    .min(0)
    .describe("Number of MegaPoints in the overall inputs.")
});
export type ReconstructionCost = z.infer<typeof ReconstructionCostSchema>;